
Привет, меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Спасибо за такую хорошую работу, мне она очень понравилась 
- Я оставил некоторые советы, надеюсь они будут полезными или интересными.
- Есть небольшие недочеты, но думаю у тебя не составить труда их исправить
- Жду обновленную работу!



<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление[3]</font>
* Спасибо за оперативность
- Приятно видеть фидбек на комментарии.
- Недочеты исправлены - работа полностью корректна
- Рад был быть полезен. Удачи в дальнейшем обучении!


# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

# Инструкция по выполнению проекта

1 Откройте файл с данными и изучите его. 
Путь к файлу: /datasets/users_behavior.csv.

2 Разделите исходные данные на обучающую, валидационную и тестовую выборки.

3 Исследуйте качество разных моделей, меняя гиперпараметры. Кратко напишите выводы исследования.

4 Проверьте качество модели на тестовой выборке.

5 Дополнительное задание: проверьте модели на вменяемость. Ничего страшного, если не получится: эти данные сложнее тех, с которыми вы работали раньше. В следующем курсе подробнее об этом расскажем.

# Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:

сalls — количество звонков,

minutes — суммарная длительность звонков в минутах,

messages — количество sms-сообщений,

mb_used — израсходованный интернет-трафик в Мб,

is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Откройте и изучите файл

<div class="alert alert-block alert-info">
<b>Совет:</b> Советую собирать все импорты в верхней части ноутбука! 
Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>


In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

df = pd.read_csv('/datasets/users_behavior.csv')
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

<div class="alert alert-block alert-info">
<b>Совет:</b> У нас тут не задача регресии - модели которые ты в первой ячейки импортируешь не нужны в этой работе.
</div>


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [3]:
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

Как и говорилось в описании проекта, перед нами полностью гоовые данные

 
<div class="alert alert-info">
<b>Совет:</b>
Также важно изучить корреляции в данных. Так мы можем оценить взаимосвязи в данных и влияние параметров на целевой признак. Рекомендую посмотреть метод `sns.pairplot` с параметром hue, в который мы передаем целевой признак. Будет здорово, если ты сможешь это реализовать в текущем или последующих проектах.
</div>


## Разбейте данные на выборки

In [4]:
#признаки
features = df.drop(['is_ultra'], axis=1)
#целевой признак
target = df['is_ultra']

#разделим данные на обучающую, валидационную и тестовую выборки, соотношение 3:1:1
features_train, features_part, target_train, target_part = train_test_split(features, target, test_size=0.25, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_part, target_part, test_size=0.5, random_state=12345)

In [5]:
#размеры новых выборок
print(features_train.shape, features_valid.shape, features_test.shape, target_train.shape,  target_valid.shape, target_test.shape)

(2410, 4) (402, 4) (402, 4) (2410,) (402,) (402,)


<div class="alert alert-block alert-success">  
<b>Успех:</b> С разбиением все в порядке, молодец, что не забываешь проверять полученные выборки ,  это иной раз может предостеречь тебя от глупых ошибок.
</div>  

## Исследуйте модели

### Дерево принятия решения

In [6]:
best_model = 0
best_result = 0
for depth in range(1, 25):
    model = DecisionTreeClassifier(random_state=12345, max_depth = depth) # обучаем модель с заданной глубиной дерева
    model.fit(features_train, target_train) # обучаем модель
    predictions_valid = model.predict(features_valid) # получаем предсказания модели
    result = accuracy_score(target_valid, predictions_valid)
    if result > best_result:
        best_model = model
        best_result = result
        
print('Accuracy лучшей модели:', best_result)

Accuracy лучшей модели: 0.7985074626865671


### Случайный лес

In [7]:
best_est = 0
best_result = 0
best_depth = 0

for estim in range(1,25):
    for depth in range(1, 25):
        model = RandomForestClassifier(random_state=12345, n_estimators=estim, criterion='entropy', max_depth = depth)
        model.fit(features_train, target_train)
        predictions_valid = model.predict(features_valid)
        accuracy = accuracy_score(target_valid, predictions_valid)
        if accuracy > best_result:
            best_est = estim
            best_result = accuracy
            best_depth = depth

print('Количество деревьев', best_est, 'Глубина:', best_depth, 'Accuracy лучшей модели:', best_result)         

Количество деревьев 7 Глубина: 10 Accuracy лучшей модели: 0.818407960199005


При использовании дополнительных гиперпараметров показатель показатель accuracy улучшился почти на 2%. Считаю, что добились хорошего результата, проверим на тестовой выборке.


<div class="alert alert-block alert-danger">
<b>Ошибка:</b> 
    
1) В чем смысл добавлять глубину, и нигде не сохранять/отслеживать/выводить лучшее значение глубины.
    
2) Я так понял, ты здесь пыталась подобрать параметр глубина, но тогда здесь не нужно делать внешний цикл  `for estim in range(1,25):` - потому что это бессмысленно.
    
3) Не нужно каждый гиперпараметр подбирать в 1 ячейке, все нужно делать в одной используя вложенные циклы
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Вроде исправил ошибки насколько я понял. В решении этого проекта я опирался на теоритические задания из курса. Открывая и повторяя пройденные задачи.
</div>


<div class="alert alert-block alert-danger">
<b>Ошибка[2]:</b> 
    
1) Вместо трех ячеек с RandomForestClassifier сделай одну (две убери).
    
2) Сделай вложенный цикл: внешним циклом пусть у тебя остается `for estim in range(1,25)`, а внутренним циклом добавь подбор depth. Не забывай значие подставлять в классификатор, по аналоги с `estim`
</div>


<div class="alert alert-block alert-success">
<b>Успех[3]:</b> Умница!
</div>


### Логистическая регрессия

In [8]:
model =  LogisticRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
accuracy = accuracy_score(target_valid, predictions)
print('Accuracy:', accuracy)

Accuracy: 0.7039800995024875


# Вывод

По результатам расчетов мы видим, что лучше всего себя показала модель Случайного леса с результатом Accuracy 0.818407960199005(почти 82% точности). Проверим эту модель на тестовой выборке. Хуже всего показала себя модель логистической регресии  


<div class="alert alert-block alert-success">
<b>Успех[3]:</b> Шаг проделан корректно, валидационная выборка корректно использована
</div>


## Проверьте модель на тестовой выборке

In [9]:
#проверяем модель случайного леса на тестовой выборке
model = RandomForestClassifier(random_state=12345, n_estimators=7, criterion='entropy', max_depth = 10)
model.fit(features_train, target_train)
predictions = model.predict(features_test)
accuracy = accuracy_score(target_test, predictions)
print('Accuracy:', accuracy)

Accuracy: 0.7985074626865671



<div class="alert alert-block alert-danger">
<b>Ошибка:</b> 
    
После исправления ошибок на предыдущем шаге. Выбери лучшие гиперпараметры полученные на предыдущем шаге.
</div>


<div class="alert alert-block alert-success">
<b>Успех[3]:</b> С тестированием все в порядке
</div>


# Вывод
В ходе исследования мы провели ряд действий: открыли и изучили данные перед началом исследования, разбив данные на признаки и целевой признак разделили их на обучающую, валидационную и тестовую выборки, опробовали разные модели и выбрали лучшую, затем проверили модель на тестовой выборке.
Проверив модель случайного леса на тестовой выборке мы получили результат accuracy 0.7985074626865671, то есть почти 80% точности, что при маленьком размере выборки является хорошим результатом. 

## (бонус) Проверьте модели на адекватность

<div class="alert alert-block alert-info">
<b>Совет:</b> Здесь подразумевалось сравнить наши модели с фиктивной/случайной, может нам стоит всегда предсказывать наибольший класс и результат будет не хуже. Для данной задачи удобен инструмент DummyClassifier</div>
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента[2]:</b> Подойдет ли модель DummyClassifier для проверки на адекватность?
</div>

<div class="alert alert-block alert-success">
    
<b>Ревьювер[2]:</b> Можешь посмотреть здесь `https://github.com/Dyakonov/IML/tree/master/2021`. Исправь крассные комментарии - там не сложно
</div>


<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Спасибо за помощь, обязательно всё прочитаю
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
